src/components/fields/image-upload/image-manager/image-manager.ts

```diff
@@ -0,0 +1,361 @@
+import { useContext, useEffect, useRef, useState } from "react";
+import { usePrevious } from "../../../../utils/hooks";
+import { AxiosApiClient, FileHelper, ImageHelper } from "../../../../utils";
+import { ImageContext } from "../image-context";
+import {
+	EImageStatus,
+	IImage,
+	TImageUploadAcceptedFileType,
+	TImageUploadOutputFileType,
+	TUploadMethod,
+} from "../types";
+
+interface IProps {
+	accepts: TImageUploadAcceptedFileType[];
+	compress: boolean;
+	dimensions: { width: number; height: number };
+	editImage: boolean;
+	onChange: (...event: any[]) => void;
+	outputType: TImageUploadOutputFileType;
+	maxSizeInKb: number;
+	upload?: {
+		method: TUploadMethod;
+		url: string;
+	};
+	value: any;
+}
+
+/**
+ * manages selected images by listening to images from context provider
+ * rename / compress / upload
+ */
+export const ImageManager = (props: IProps) => {
+	// =============================================================================
+	// CONST, STATE, REFS
+	// =============================================================================
+	const { accepts, compress, dimensions, editImage, maxSizeInKb, onChange, outputType, upload, value } = props;
+	const { images, setImages, setErrorCount } = useContext(ImageContext);
+	const previousImages = usePrevious(images);
+	const [managerErrorCount, setManagerErrorCount] = useState(0);
+	const previousValue = usePrevious(value);
+	const sessionId = useRef<string>();
+
+	// =============================================================================
+	// EFFECTS
+	// =============================================================================
+	// generate pseudo-random session id
+	useEffect(() => {
+		sessionId.current = Array(5)
+			.fill(0)
+			.map(() => Math.random().toString(36).slice(2))
+			.join("");
+	}, []);
+
+	useEffect(() => {
+		images.forEach((image, index) => {
+			const previousFile = previousImages?.[index];
+			if (image.status !== previousFile?.status || image.dataURL !== previousFile.dataURL) {
+				switch (image.status) {
+					case EImageStatus.INJECTED:
+						FileHelper.dataUrlToBlob(image.dataURL)
+							.then((blob) => {
+								setImages((prev) => {
```

Comment:
is it worth using a library like `immer` to handle immutable updates?

Reply:
thanks for the suggestion, `immer` is a good library and can clean up the codes a bit, however I am trying to reduce the amount of external dependencies hence I am not intending to use it just yet, I am looking to use it when we face more scenarios that require immutable updates.